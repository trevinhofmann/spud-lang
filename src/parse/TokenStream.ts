import type { Token, TokenType } from '../types/tokens';

type Parser<T> = (tokenStream: TokenStream) => T;

export default class TokenStream {
  private readonly tokens: Token[];

  private index: number = 0;

  public constructor(tokens: Token[]) {
    this.tokens = tokens;
  }

  public peekToken = (index = 0): Token => {
    if (!this.hasTokens(index + 1)) {
      throw new Error('Token expected');
    }
    index += this.index;
    return this.tokens[index];
  };

  public peekTokenType = (type: TokenType, index = 0): Token => {
    const token = this.peekToken(index);
    if (token.type !== type) {
      throw this.error(`Expected ${type} token`);
    }
    return token;
  };

  public peekTokenTypes = (types: TokenType[]): Token[] => (
    types.map((type, i) => this.peekTokenType(type, i))
  );

  public readToken = (): Token => {
    if (!this.hasTokens(1)) {
      throw new Error('Token expected');
    }
    const token = this.tokens[this.index];
    this.index += 1;
    return token;
  };

  public readTokenType = (type: TokenType): Token => {
    const token = this.readToken();
    if (token.type !== type) {
      throw this.error(`Expected "${type}" token`);
    }
    return token;
  };

  public readTokenTypes = (types: TokenType[]): Token[] => (
    types.map(type => this.readTokenType(type))
  );

  public readAllOfTokenType = (type: TokenType, minimum = 1): Token[] => {
    const tokens: Token[] = [];
    while (this.checkTokenType(type)) {
      tokens.push(this.readTokenType(type));
    }
    if (tokens.length < minimum) {
      throw new Error(`Expected at least ${minimum} ${type} tokens. Found ${tokens.length}.`);
    }
    return tokens;
  };

  public checkTokenType = (type: TokenType, index = 0): boolean => {
    if (!this.hasTokens()) {
      return false;
    }
    const token = this.peekToken(index);
    return token.type === type;
  };

  public checkTokenTypes = (types: TokenType[]): boolean => {
    if (!this.hasTokens(types.length)) {
      return false;
    }
    return types.reduce((matches, type, i) => (
      matches && this.checkTokenType(type, i)
    ), true);
  };

  public hasTokens = (quantity = 1): boolean => {
    quantity += this.index;
    return this.tokens.length >= quantity;
  };

  public peek = <T>(parser: Parser<T>, index = 0): T => {
    const originalIndex = this.index;
    this.index += index;
    const node = parser(this);
    this.index = originalIndex;
    return node;
  };

  public read = <T>(parser: Parser<T>): T => {
    const node = parser(this);
    return node;
  };

  public readAll = <T>(parser: Parser<T>): T[] => {
    const nodes: T[] = [];
    while (this.has(parser)) {
      nodes.push(this.read(parser));
    }
    return nodes;
  };

  public has = <T>(parser: Parser<T>, index = 0): boolean => {
    const originalIndex = this.index;
    try {
      this.peek(parser, index);
      return true;
    } catch (e: unknown) {
      return false;
    } finally {
      this.index = originalIndex;
    }
  };

  public error = (message: string, token?: Token): Error => {
    if (token === undefined) {
      token = this.peekToken();
    }
    message = `${message}.
Token: ${token.type} (${token.characters})
Location: Line ${token.line}, Column ${token.colStart}`;
    return new Error(message);
  };
}
