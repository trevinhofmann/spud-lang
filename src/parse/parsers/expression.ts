import type { Expression } from '../../types/ast';
import type TokenStream from '../TokenStream';
import binaryExpression from './binaryExpression';
import subbinaryExpression from './subbinaryExpression';

export default (tokenStream: TokenStream): Expression => {
  if (tokenStream.has(binaryExpression)) {
    const result = tokenStream.read(binaryExpression);
    return result;
  }
  if (tokenStream.has(subbinaryExpression)) {
    return tokenStream.read(subbinaryExpression);
  }
  throw tokenStream.error('Expeced an expression');
};
