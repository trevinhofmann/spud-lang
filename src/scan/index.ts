import Logger from 'eleventh';
import { TokenType } from '../types/tokens';
import type { Token, PartialToken } from '../types/tokens';
import symbols from '../utils/symbols';
import reservedWords from '../utils/reservedWords';

const commentRegex = /^\/\/.*/u;
const integerRegex = /^[0-9]+/u;
const stringRegex = /^'[^']*'/u;
const wordRegex = /^[_a-zA-Z][_a-zA-Z0-9]*/u;

type Scanner = (remainingText: string) => PartialToken | undefined;

const commentScanner: Scanner = (remainingText) => {
  const matches = commentRegex.exec(remainingText);
  if (matches !== null) {
    return {
      type: TokenType.Comment,
      characters: matches[0],
    };
  }
};

const symbolScanner: Scanner = (remainingText) => {
  for (const { characters, type } of symbols) {
    if (remainingText.startsWith(characters)) {
      return {
        type,
        characters,
      };
    }
  }
};

const integerScanner: Scanner = (remainingText) => {
  const matches = integerRegex.exec(remainingText);
  if (matches !== null) {
    return {
      type: TokenType.IntegerLiteral,
      characters: matches[0],
    };
  }
};

const stringScanner: Scanner = (remainingText) => {
  const matches = stringRegex.exec(remainingText);
  if (matches !== null) {
    return {
      type: TokenType.StringLiteral,
      characters: matches[0],
    };
  }
};

const wordScanner: Scanner = (remainingText) => {
  const matches = wordRegex.exec(remainingText);
  if (matches !== null) {
    const characters = matches[0];
    const reservedWord = reservedWords
      .find(word => word.characters === characters);
    if (reservedWord !== undefined) {
      return {
        type: reservedWord.type,
        characters,
      };
    }
    return {
      type: TokenType.UserDefinedWord,
      characters,
    };
  }
};

const scanners = [
  commentScanner,
  symbolScanner,
  integerScanner,
  stringScanner,
  wordScanner,
];

export default (characters: string): Token[] => {
  let lineNumber = 1;
  let i = 0;
  let remainingText = characters;
  const tokens: Token[] = [];
  while (remainingText.length > 0) {
    const partialToken = scanners.reduce<PartialToken | undefined>(
      (token, scanner) => token ?? scanner(remainingText),
      undefined,
    );
    if (partialToken === undefined) {
      throw new Error(`Unable to scan token at ${remainingText}`);
    }
    Logger.info('Scanned token', { tokenType: partialToken.type, characters: partialToken.characters });
    tokens.push({
      ...partialToken,
      line: lineNumber,
      colStart: i,
      colEnd: i + partialToken.characters.length,
    });
    if (partialToken.type === TokenType.Newline) {
      i = 0;
      lineNumber += 1;
    } else {
      i += partialToken.characters.length;
    }
    remainingText = remainingText.slice(partialToken.characters.length);
  }
  return tokens;
};
