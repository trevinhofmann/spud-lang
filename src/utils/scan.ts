import Logger from 'eleventh';
import { TokenType } from '../types/tokens';
import type { Line } from '../types/lines';
import type { Token, PartialToken } from '../types/tokens';
import symbols from './symbols';
import reservedWords from './reservedWords';

const commentRegex = /^\/\/.*/u;
const integerRegex = /^[0-9]+/u;
const wordRegex = /^[_a-zA-Z][_a-zA-Z0-9]*/u;

type Scanner = (remainingText: string) => PartialToken | undefined;

const commentScanner: Scanner = (remainingText) => {
  const matches = commentRegex.exec(remainingText);
  if (matches !== null) {
    return {
      type: TokenType.Comment,
      characters: matches[0],
    };
  }
};

const symbolScanner: Scanner = (remainingText) => {
  for (const { characters, type } of symbols) {
    if (remainingText.startsWith(characters)) {
      return {
        type,
        characters,
      };
    }
  }
};

const integerScanner: Scanner = (remainingText) => {
  const matches = integerRegex.exec(remainingText);
  if (matches !== null) {
    return {
      type: TokenType.Integer,
      characters: matches[0],
    };
  }
};

const wordScanner: Scanner = (remainingText) => {
  const matches = wordRegex.exec(remainingText);
  if (matches !== null) {
    const characters = matches[0];
    const reservedWord = reservedWords
      .find(word => word.characters === characters);
    if (reservedWord !== undefined) {
      return {
        type: reservedWord.type,
        characters,
      };
    }
    return {
      type: TokenType.UserDefinedWord,
      characters,
    };
  }
};

const scanners = [
  commentScanner,
  symbolScanner,
  integerScanner,
  wordScanner,
];

const scanLine = (characters: string, lineNumber: number): Line => {
  let i = 0;
  let remainingText = characters;
  const tokens: Token[] = [];
  const line: Line = {
    lineNumber,
    characters,
    tokens,
  };
  while (i < characters.length) {
    const partialToken = scanners.reduce<PartialToken | undefined>(
      (token, scanner) => token ?? scanner(remainingText),
      undefined,
    );
    if (partialToken === undefined) {
      throw new Error('Unable to scan token');
    }
    Logger.info('Scanned token', { tokenType: partialToken.type, characters: partialToken.characters });
    tokens.push({
      ...partialToken,
      line,
      colStart: i,
      colEnd: i + partialToken.characters.length,
    });
    i += partialToken.characters.length;
    remainingText = remainingText.slice(partialToken.characters.length);
  }
  return line;
};

export const scanLines = (file: string): Line[] => (
  file
    .split('\n')
    .map((characters, i) => scanLine(characters, i + 1))
);
