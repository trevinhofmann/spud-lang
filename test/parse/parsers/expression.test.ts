import scan from '../../../src/scan';
import expression from '../../../src/parse/parsers/expression';
import TokenStream from '../../../src/parse/TokenStream';

const testCases = [
  '4 + 5',
  'a + b',
  '1 + 2 + 3',
  '4 * 7 + a * b',
  '!a',
  '-b',
];

describe('expression', () => {
  testCases.forEach(testCase => {
    it(testCase, () => {
      const tokens = scan(testCase);
      const tokenStream = new TokenStream(tokens);
      const result = tokenStream.read(expression);
      expect(result).toMatchSnapshot();
    });
  });
});