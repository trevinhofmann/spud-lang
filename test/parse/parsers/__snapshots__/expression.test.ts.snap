// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expression !a 1`] = `
{
  "operand": {
    "name": "a",
    "type": "UntypedVariable",
  },
  "operator": "Negation",
  "type": "UnaryExpression",
}
`;

exports[`expression -b 1`] = `
{
  "operand": {
    "name": "b",
    "type": "UntypedVariable",
  },
  "operator": "Negative",
  "type": "UnaryExpression",
}
`;

exports[`expression 1 + 2 + 3 1`] = `
{
  "left": {
    "left": {
      "type": "NumericLiteral",
      "value": "1",
    },
    "operator": "Addition",
    "right": {
      "type": "NumericLiteral",
      "value": "2",
    },
    "type": "BinaryExpression",
  },
  "operator": "Addition",
  "right": {
    "type": "NumericLiteral",
    "value": "3",
  },
  "type": "BinaryExpression",
}
`;

exports[`expression 4 * 7 + a * b 1`] = `
{
  "left": {
    "left": {
      "type": "NumericLiteral",
      "value": "4",
    },
    "operator": "Multiplication",
    "right": {
      "type": "NumericLiteral",
      "value": "7",
    },
    "type": "BinaryExpression",
  },
  "operator": "Addition",
  "right": {
    "left": {
      "name": "a",
      "type": "UntypedVariable",
    },
    "operator": "Multiplication",
    "right": {
      "name": "b",
      "type": "UntypedVariable",
    },
    "type": "BinaryExpression",
  },
  "type": "BinaryExpression",
}
`;

exports[`expression 4 + 5 1`] = `
{
  "left": {
    "type": "NumericLiteral",
    "value": "4",
  },
  "operator": "Addition",
  "right": {
    "type": "NumericLiteral",
    "value": "5",
  },
  "type": "BinaryExpression",
}
`;

exports[`expression a + b 1`] = `
{
  "left": {
    "name": "a",
    "type": "UntypedVariable",
  },
  "operator": "Addition",
  "right": {
    "name": "b",
    "type": "UntypedVariable",
  },
  "type": "BinaryExpression",
}
`;
